src/btree/bt_sync.c

```diff
@@ -414,9 +414,17 @@ __wt_sync_file(WT_SESSION_IMPL *session, WT_CACHE_OP syncop)
     if (txn->isolation == WT_ISO_READ_COMMITTED && saved_pinned_id == WT_TXN_NONE)
         __wt_txn_release_snapshot(session);
 
+    /*
+     * Ensure the checkpoint generation is updated before clearing the sync flag. Otherwise,
+     * eviction could evict a page from the btree after the flag is cleared but before the
+     * checkpoint generation is updated. This would violate the constraints of disaggregated
+     * storage, as eviction would write a page that should not be part of the current checkpoint.
+     */
+    __wt_checkpoint_update_generation(session, btree);
+
     /* Clear the checkpoint flag. */
-    __wt_atomic_store_enum_relaxed(&btree->syncing, WT_BTREE_SYNC_OFF);
-    __wt_atomic_store_ptr_relaxed(&btree->sync_session, NULL);
+    __wt_atomic_store_enum_release(&btree->syncing, WT_BTREE_SYNC_OFF);
+    __wt_atomic_store_ptr_release(&btree->sync_session, NULL);
```

Comment:
I tried to understand what memory are getting protected by storing all these variables (`btree->checkpoint_gen`, `btree->syncing` and `btree->sync_session`) with release ordering. Is that protection from these operations being reordered with each other?

I one atomic store with release should protect us from reordering with another one, then I'll prefer having one that is getting protected as relaxed.

Reply:
But why do we need acquire then?

> Theoretically, these should be sequential consistency here as we need synchronization across variables.

What variables are we synchronising here?

Reply:
the sync_session read itself

We now do:

```
#define WT_BTREE_SYNCING(btree) \
    (__wt_atomic_load_enum_acquire(&(btree)->syncing) != WT_BTREE_SYNC_OFF)
#define WT_SESSION_BTREE_SYNC(session) \
    (__wt_atomic_load_ptr_acquire(&S2BT(session)->sync_session) == (session))
#define WT_SESSION_BTREE_SYNC_SAFE(session, btree)                                \
    (__wt_atomic_load_enum_acquire(&(btree)->syncing) != WT_BTREE_SYNC_RUNNING || \
      __wt_atomic_load_ptr_acquire(&(btree)->sync_session) == (session))
```

Reply:
Yeah, I see, but I still don't get it, maybe it's worth discussing this on a call. I mean what would break if we put relaxed everywhere? Why do we need acquire + release in all three places?

Reply:
These variables are read by different threads and there is a requirement for them to synchronize. Therefore, they should use acquire + release.